pkg/virt-operator/kubevirt.go

```diff
@@ -0,0 +1,362 @@
+	"kubevirt.io/kubevirt/pkg/virt-operator/deletion"
+	"kubevirt.io/kubevirt/pkg/virt-operator/util"
+)
+
+const (
+	ConditionReasonDeploymentFailedExisting = "ExistingDeployment"
+	ConditionReasonDeploymentFailedError    = "DeploymentFailed"
+	ConditionReasonDeletionFailedError      = "DeletionFailed"
+)
+
+type KubeVirtController struct {
+	clientset            kubecli.KubevirtClient
+	queue                workqueue.RateLimitingInterface
+	kubeVirtInformer     cache.SharedIndexInformer
+	recorder             record.EventRecorder
+	config               util.KubeVirtDeploymentConfig
+	stores               util.Stores
+	informers            []cache.SharedIndexInformer
+	kubeVirtExpectations *controller.ControllerExpectations
+}
+
+func NewKubeVirtController(
+	clientset kubecli.KubevirtClient,
+	informer cache.SharedIndexInformer,
+	recorder record.EventRecorder,
+	stores util.Stores,
+	informers []cache.SharedIndexInformer) *KubeVirtController {
+
+	c := KubeVirtController{
+		clientset:            clientset,
+		queue:                workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		kubeVirtInformer:     informer,
+		recorder:             recorder,
+		config:               util.GetConfig(),
+		stores:               stores,
+		informers:            informers,
+		kubeVirtExpectations: controller.NewControllerExpectations(),
+	}
+
+	c.kubeVirtInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addKubeVirt,
+		DeleteFunc: c.deleteKubeVirt,
+		UpdateFunc: c.updateKubeVirt,
+	})
+
+	for _, informer := range informers {
+
+		informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addHandler,
+			DeleteFunc: c.deleteHandler,
+			UpdateFunc: c.updateHandler,
+		})
+	}
+
+	return &c
+}
+
+func (c *KubeVirtController) genericHandler(obj interface{}, isCreate bool) {
+	logger := log.Log
+	var object metav1.Object
+	var ok bool
+
+	if object, ok = obj.(metav1.Object); !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			return
+		}
+		logger.V(4).Infof("Recovered deleted object '%s' from tombstone", object.GetName())
+	}
+
+	logger.V(4).Infof("Observed object %s in handler. isCreate: %t", object.GetName(), isCreate)
+
+	// add/delete detected... enqueue active kubevirt objects.
+	kvs := c.kubeVirtInformer.GetStore().List()
+	for _, obj := range kvs {
+		if kv, ok := obj.(v1.KubeVirt); ok {
+			if isKubeVirtActive(&kv) {
+				key, err := controller.KeyFunc(kv)
+				if err != nil {
+					logger.Object(&kv).Reason(err).Error("Failed to extract key from KubeVirt.")
+				}
+
+				if isCreate {
+					c.kubeVirtExpectations.CreationObserved(key)
+				} else {
+					c.kubeVirtExpectations.DeletionObserved(key)
+				}
+				c.enqueueKubeVirt(obj)
+			}
+		}
+	}
+}
+
+func (c *KubeVirtController) addHandler(obj interface{}) {
+	c.genericHandler(obj, true)
```

Comment:
I am not sure if it will cause issues right now, but normally `add` callbacks need to check if the deletionTimestamp is set, to avoid counting deleted objects as an observed creation if the controller is restarted.